nautobot/utilities/filters.py

```diff
@@ -286,6 +286,141 @@ def filter(self, qs, value):
+        )
+
+    Optionally you may also provide a callable to use as a preprocessor for the filter predicate by
+    providing the value as a nested dict with "lookup_expr" and "preprocessor" keys. For example:
+
+        q = SearchFilter(
+            filter_predicates={
+                "asn": {
+                    "lookup_expr": "exact",
+                    "preprocessor": int,
+            },
+        )
+
+    This tells the filter to try to cast `asn` to an `int`. If it fails, this predicate will be
+    skipped.
+    """
+
+    # Optional label for the form element generated for this filter
+    label = None
+
+    def __init__(self, filter_predicates=None, label=None, *args, **kwargs):
+        if not isinstance(filter_predicates, dict):
+            raise TypeError("filter_predicates must be a dict")
+
+        # Format: {field_name: lookup_expr, ...}
+        self.filter_predicates = filter_predicates
+
+        # Try to use the label from the class if it is defined.
+        if label is None:
+            kwargs.setdefault("label", self.label)
+
+        super().__init__(*args, **kwargs)
+
+        # Generate the query with a sentinel value to validate it and surface parse errors.
+        self.generate_query(self.filter_predicates, value="")
+
+    def generate_query(self, filter_predicates, value):
+        """
+        Given a mapping of `filter_predicates` and a `value`, return a `Q` object for 2-tuple of
+        predicate=value.
+        """
+        query = models.Q()
+        for field_name, lookup_info in filter_predicates.items():
+            if isinstance(lookup_info, str):
+                lookup_expr = lookup_info
+                preprocessor = str.strip
+            elif isinstance(lookup_info, dict):
+                lookup_expr = lookup_info["lookup_expr"]
+                preprocessor = lookup_info.get("preprocessor")
+                if not callable(preprocessor):
+                    raise TypeError("Preprocessor {preprocessor} must be callable!")
+            else:
+                raise TypeError(f"Predicate value must be a str or a dict! Got: {type(lookup_info)}")
+
+            # Try to preprocess the value or skip creating a predicate for it. In the event we try
+            # to cast a value to an invalid type (e.g. `int("foo")` or `dict(42)`), ensure this
+            # predicate is not included in the query.
+            try:
+                value = preprocessor(value)
+            except (TypeError, ValueError):
+                continue
+
+            predicate = {f"{field_name}__{lookup_expr}": value}
+            query |= models.Q(**predicate)
+
+        # Return this for later use (such as introspection or debugging)
+        return query
+
+    def filter(self, qs, value):
+        if value in EMPTY_VALUES:
+            return qs
+
+        # Evaluate the query and stash it for later use (such as introspection or debugging)
+        self.query = self.generate_query(self.filter_predicates, value)
```

Comment:
Maybe call it something like `self._most_recent_query` or similar to make it more obvious that this is *not* a general-purpose query that can be reused later, but just a cached value?